File: src/joy_mux_controller/joy_mux_controller/IK_mux.py
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Joy, JointState
from geometry_msgs.msg import TwistStamped, Twist
from control_msgs.msg import JointJog
from std_msgs.msg import Float32

#Mapping Joystick button binds:
X = 0
SQUARE = 3
TRIANGLE = 2
CIRCLE = 1
RBUMPER = 5

#Mapping Axis binds: 

LEFT_STICK_X = 0
LEFT_STICK_Y = 1
RIGHT_STICK_X = 3
RIGHT_STICK_Y = 4
D_PAD_X = 6
D_PAD_Y = 7
LTRIGGER = 2
RTRIGGER = 5

AXIS_DEFAULTS = 1.0 # Default value for triggers when not pressed

def convertJoytoCmd(joy_msg: Joy):
   
    if joy_msg.buttons[X] or joy_msg.buttons[SQUARE] or joy_msg.buttons[TRIANGLE] or joy_msg.buttons[CIRCLE] or joy_msg.axes[D_PAD_X] or joy_msg.axes[D_PAD_Y]:
        # If any of the buttons or D-Pad axes are pressed, we assume it's a joint jog command
        # and not a twist command.

        joint = JointJog()
        joint.joint_names = [
        'base_pivot_shoulder_gearbox_joint',
        'bicep_tube_gearbox_joint',
        'forearm_tube_wrist_gearbox_joint',
        'gripper_claw_joint'
        ]
        joint.velocities = [
            float(joy_msg.axes[D_PAD_X]), 
            float(joy_msg.axes[D_PAD_Y]),
            float(joy_msg.buttons[X] - joy_msg.buttons[TRIANGLE]),
            float(joy_msg.buttons[SQUARE] - joy_msg.buttons[CIRCLE])
            ]
        return True, None, joint 
    #Setting this flag to False tells the node not to publish a joint jog commands.
    #This is because were returning a 3-tuple, and the publish flag is associated to either True/False
        
    twist_stamped = TwistStamped()
    twist_stamped.twist.linear.z = joy_msg.axes[LEFT_STICK_Y]
    twist_stamped.twist.linear.y = joy_msg.axes[LEFT_STICK_X]
    
    lin_x_right = -0.5*(joy_msg.axes[RTRIGGER] - AXIS_DEFAULTS)
    lin_x_left = 0.5*(joy_msg.axes[LTRIGGER] - AXIS_DEFAULTS)
    twist_stamped.twist.linear.x = lin_x_right + lin_x_left

    return True, twist_stamped , None


class JoyMuxController(Node):
    def __init__(self):
        super().__init__('joy_mux_controller')
        self.subscription = self.create_subscription(Joy, '/joy', self.joy_callback, 10)
        self.rover_pub = self.create_publisher(Twist, '/cmd_vel', 10)
        self.arm_pub = self.create_publisher(TwistStamped, '/servo_node/delta_twist_cmds', 10) 
        self.jog_pub = self.create_publisher(JointJog, '/servo_node/delta_joint_cmds', 10)

        self.deadman_button = 4
        self.toggle_button = 10
        self.current_mode = 0
        self.last_toggle = 0

    def joy_callback(self, msg: Joy):
        
        if msg.buttons[self.toggle_button] == 1 and self.last_toggle == 0:
            self.current_mode = 1 - self.current_mode
            self.get_logger().info(f"Switched to {'Arm' if self.current_mode else 'Rover'} mode")
        self.last_toggle = msg.buttons[self.toggle_button]

        if msg.buttons[self.deadman_button] == 1:
            if self.current_mode == 0:
                twist = Twist()
                twist.linear.x = msg.axes[0]
                twist.angular.z = msg.axes[1]
                twist.linear.y = msg.axes[7]
                twist.linear.z = msg.axes[6]
                self.rover_pub.publish(twist)
            else:
                publish, twist_stamped, joint = convertJoytoCmd(msg)
                if publish and twist_stamped is not None:
                    twist_stamped.header.stamp = self.get_clock().now().to_msg()
                    twist_stamped.header.frame_id = 'base_structure_link'
                    self.arm_pub.publish(twist_stamped)
                elif publish and joint is not None:
                    joint.header.stamp = self.get_clock().now().to_msg()
                    joint.header.frame_id = 'base_structure_link'
                    self.jog_pub.publish(joint)     
        return

def main(args=None):
    rclpy.init(args=args)
    node = JoyMuxController()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/inverse-kinematics/ik_mux_controller/src/ik_mux.cpp

#include "rclcpp/rclcpp.hpp"
#include <sensor_msgs/msg/joy.hpp>
#include <geometry_msgs/msg/twist_stamped.hpp>
#include <control_msgs/msg/joint_jog.hpp>
#include <std_srvs/srv/trigger.hpp>
#include <moveit_msgs/msg/planning_scene.hpp>
#include <rclcpp/client.hpp>
#include <rclcpp/experimental/buffers/intra_process_buffer.hpp>
#include <rclcpp/node.hpp>
#include <rclcpp/publisher.hpp>
#include <rclcpp/qos.hpp>
#include <rclcpp/qos_event.hpp>
#include <rclcpp/subscription.hpp>
#include <rclcpp/time.hpp>
#include <rclcpp/utilities.hpp>
#include <thread>

// We'll just set up parameters here
const std::string JOY_TOPIC = "/joy";
const std::string TWIST_TOPIC = "/servo_node/delta_twist_cmds";
const std::string JOINT_TOPIC = "/servo_node/delta_joint_cmds";
const std::string WHEEL_VEL_TOPIC = "/cmd_vel";
const std::string EEF_FRAME_ID = "gripper_claw_link";
const std::string BASE_FRAME_ID = "base_structure_link";

// Enums for button names -> axis/button array index
// For XBOX 1 controller
enum Axis
{
  LEFT_STICK_X = 0,
  LEFT_STICK_Y = 1,
  LEFT_TRIGGER = 2,
  RIGHT_STICK_X = 3,
  RIGHT_STICK_Y = 4,
  RIGHT_TRIGGER = 5,
  D_PAD_X = 6,
  D_PAD_Y = 7
};
enum Button
{
  X = 0,
  CIRCLE = 1,
  TRIANGLE = 2,
  SQUARE = 3,
  LEFT_BUMPER = 4,
  RIGHT_BUMPER = 5,
  CHANGE_VIEW = 6,
  SHARE = 8,
  HOME = 10,
  LEFT_STICK_CLICK = 11,
  RIGHT_STICK_CLICK = 12
};

// Some axes have offsets (e.g. the default trigger position is 1.0 not 0)
// This will map the default values for the axes
std::map<Axis, double> AXIS_DEFAULTS = { { LEFT_TRIGGER, 1.0 }, { RIGHT_TRIGGER, 1.0 } };
std::map<Button, double> BUTTON_DEFAULTS;

// To change controls or setup a new controller, all you should to do is change the above enums and the follow 2
// functions
/** \brief // This converts a joystick axes and buttons array to a TwistStamped or JointJog message
 * @param axes The vector of continuous controller joystick axes
 * @param buttons The vector of discrete controller button values
 * @param twist A TwistStamped message to update in prep for publishing
 * @param joint A JointJog message to update in prep for publishing
 * @return return true if you want to publish a Twist, false if you want to publish a JointJog
 */
bool convertJoyToCmd(const std::vector<float>& axes, const std::vector<int>& buttons,
                     std::unique_ptr<geometry_msgs::msg::TwistStamped>& twist,
                     std::unique_ptr<control_msgs::msg::JointJog>& joint)
{
  // Give joint jogging priority because it is only buttons
  // If any joint jog command is requested, we are only publishing joint commands
  if (buttons[X] || buttons[CIRCLE] || buttons[TRIANGLE] || buttons[SQUARE] || axes[D_PAD_X] || axes[D_PAD_Y] || buttons[RIGHT_STICK_CLICK] || buttons[LEFT_STICK_CLICK])
  {
    // Map the D_PAD to the proximal joints
    joint->joint_names.push_back("joint1");
    joint->velocities.push_back(axes[D_PAD_X]);
    joint->joint_names.push_back("joint2");
    joint->velocities.push_back(axes[D_PAD_Y]);

    // Map the diamond to the distal joints
    joint->joint_names.push_back("joint3");
    joint->velocities.push_back(buttons[CIRCLE] - buttons[SQUARE]);
    joint->joint_names.push_back("joint5");
    joint->velocities.push_back(buttons[TRIANGLE] - buttons[X]);
    // joint->joint_names.push_back("joint7");
    // joint->velocities.push_back(buttons[LEFT_STICK_CLICK] - buttons[RIGHT_STICK_CLICK]);
    return false;
  }

  // The bread and butter: map buttons to twist commands
  twist->twist.linear.z = axes[RIGHT_STICK_X];
  twist->twist.linear.y = axes[RIGHT_STICK_Y];

  double lin_x_right = -0.5 * (axes[RIGHT_TRIGGER] - AXIS_DEFAULTS.at(RIGHT_TRIGGER));
  double lin_x_left = 0.5 * (axes[LEFT_TRIGGER] - AXIS_DEFAULTS.at(LEFT_TRIGGER));
  twist->twist.linear.x = lin_x_right + lin_x_left;

  twist->twist.angular.y = axes[LEFT_STICK_Y];
  twist->twist.angular.x = axes[LEFT_STICK_X];

  double roll_positive = buttons[RIGHT_BUMPER];
  double roll_negative = -1 * (buttons[LEFT_BUMPER]);
  twist->twist.angular.z = roll_positive + roll_negative;

  return true;
}


namespace moveit_servo
{
class JoyToServoPub : public rclcpp::Node
{
public:
  JoyToServoPub(const rclcpp::NodeOptions& options)
    : Node("joy_to_twist_publisher", options), frame_to_publish_(BASE_FRAME_ID)
  {
    // Setup pub/sub
    joy_sub_ = this->create_subscription<sensor_msgs::msg::Joy>(
        JOY_TOPIC, rclcpp::SystemDefaultsQoS(),
        [this](const sensor_msgs::msg::Joy::ConstSharedPtr& msg) { return joyCB(msg); });

    twist_pub_ = this->create_publisher<geometry_msgs::msg::TwistStamped>(TWIST_TOPIC, rclcpp::SystemDefaultsQoS());
    joint_pub_ = this->create_publisher<control_msgs::msg::JointJog>(JOINT_TOPIC, rclcpp::SystemDefaultsQoS());
    collision_pub_ =
        this->create_publisher<moveit_msgs::msg::PlanningScene>("/planning_scene", rclcpp::SystemDefaultsQoS());

    // Create a service client to start the ServoNode
    servo_start_client_ = this->create_client<std_srvs::srv::Trigger>("/servo_node/start_servo");
    servo_start_client_->wait_for_service(std::chrono::seconds(1));
    servo_start_client_->async_send_request(std::make_shared<std_srvs::srv::Trigger::Request>());
    RCLCPP_INFO(this->get_logger(), "Ik mux node publishing topic data");

  }

  ~JoyToServoPub() override
  {
    if (collision_pub_thread_.joinable())
      collision_pub_thread_.join();
  }

  void joyCB(const sensor_msgs::msg::Joy::ConstSharedPtr& msg)
  {
    // Create the messages we might publish
    auto twist_msg = std::make_unique<geometry_msgs::msg::TwistStamped>();
    auto joint_msg = std::make_unique<control_msgs::msg::JointJog>();

      // Convert the joystick message to Twist or JointJog and publish
      if (convertJoyToCmd(msg->axes, msg->buttons, twist_msg, joint_msg))
      {
        // publish the TwistStamped
        twist_msg->header.frame_id = frame_to_publish_;
        twist_msg->header.stamp = this->now();
        twist_pub_->publish(std::move(twist_msg));
      }
      else
      {
        // publish the JointJog
        joint_msg->header.stamp = this->now();
        joint_msg->header.frame_id = "base_structure_link";
        joint_msg->duration = 1.0;
        joint_pub_->publish(std::move(joint_msg));
      }
    
  }

private:
  rclcpp::Subscription<sensor_msgs::msg::Joy>::SharedPtr joy_sub_;
  rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr twist_pub_;
  rclcpp::Publisher<control_msgs::msg::JointJog>::SharedPtr joint_pub_;
  rclcpp::Publisher<moveit_msgs::msg::PlanningScene>::SharedPtr collision_pub_;
  rclcpp::Client<std_srvs::srv::Trigger>::SharedPtr servo_start_client_;
  rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr wheel_pub_; 
  std::string frame_to_publish_;

  std::thread collision_pub_thread_;
};  // class JoyToServoPub

}  // namespace moveit_servo

// Register the component with class_loader
int main(int argc, char** argv)
{
  rclcpp::init(argc,argv);
  auto node = std::make_shared<moveit_servo::JoyToServoPub>(rclcpp::NodeOptions{});
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}